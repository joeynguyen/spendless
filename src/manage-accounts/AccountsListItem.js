import React, { Component, PropTypes } from 'react';
import { reduxForm } from 'redux-form';
import { bindActionCreators } from 'redux';
import { Well, Collapse, Input, Button, Alert } from 'react-bootstrap';
import { toggleAccountDeletedConfirm, storeDeletedAccountName } from '../manage-accounts/ManageAccountsActions.js';

// PouchDB is loaded externally through a script tag in the browser
const db = new PouchDB('accounts');

class AccountsListItem extends Component {
  static propTypes = {
    account: PropTypes.object.isRequired,
    fields: PropTypes.object.isRequired,
    resetForm: PropTypes.func.isRequired,
    pristine: PropTypes.bool.isRequired,
    doToggleAccountDeletedConfirm: PropTypes.func.isRequired,
    doStoreDeletedAccountName: PropTypes.func.isRequired,
  }
  state = {
    settingsVisible: false,
    alertVisible: false,
    confirmDeleteVisible: false,
    confirmDeleteText: '',
  }
  toggleSettings = () => {
    if (this.state.settingsVisible) {
      this.props.resetForm();
    }
    this.setState({ settingsVisible: !this.state.settingsVisible });
  }
  toggleConfirmDelete = () => {
    this.setState({ confirmDeleteVisible: !this.state.confirmDeleteVisible });
  }
  handleConfirmDeleteText = (e) => {
    this.setState({ confirmDeleteText: e.target.value });
  }
  handleUpdateAccount = (accountToUpdate) => {
    const self = this;
    const newAccountObj = Object.assign({}, accountToUpdate, {
      name: this.props.fields.accountName.value,
      type: this.props.fields.accountType.value,
      company: this.props.fields.accountCompany.value,
    });
    // Update account in DB
    db.put(newAccountObj).then(function(result) {
      console.log('Successfully updated account', result);
      self.setState({alertVisible: true});
    }).catch(function(err) {
      console.log(err);
      // TODO: Add error message after update fail
    });
  }
  handleDeleteAccount = (accountToDelete) => {
    const self = this;
    // Remove account from DB
    db.remove(accountToDelete).then(function(result) {
      console.log('Successfully deleted account', result);
      self.props.doStoreDeletedAccountName(accountToDelete.name);
    }).then(function() {
      self.props.doToggleAccountDeletedConfirm();
    }).catch(function(err) {
      console.log('Error trying to delete account', err);
      // TODO: Add error message after delete fail
    });
  }

  render() {
    const styles = {
      removeIcon: {
        cursor: 'pointer',
      }
    };
    const { fields: { accountName, accountType, accountCompany } } = this.props;
    let alertMessage;

    if (this.state.alertVisible) {
      alertMessage = (
        <div>
          <br />
          <Alert bsStyle="success" onDismiss={() => this.setState({alertVisible: false})} dismissAfter={2000}>
            <p>Account updated successfully!</p>
          </Alert>
        </div>
      );
    }

    return (
      <Well bsSize="small" key={this.props.account._id}>
        <i className="fa fa-lg fa-fw fa-cog pull-right" style={styles.removeIcon} onClick={this.toggleSettings} ></i>
        <h4>{this.props.account.name}</h4>
        <p>{this.props.account.type === 'bank' ? 'Bank' : 'Credit Card'} - {this.props.account.company}</p>
        <Collapse in={this.state.settingsVisible}>
          <div>
            <form>
              <Input
                type="text"
                label="Name"
                {...accountName}
              />
              <Input
                type="select"
                label="Type"
                {...accountType}
              >
                <option value="">select...</option>
                <option value="bank">Bank</option>
                <option value="creditcard">Credit Card</option>
              </Input>
              <Collapse in={accountType.value === 'bank'}>
                <div>
                  <Input
                    type="text"
                    label="Name of Institution"
                    {...accountCompany}
                  />
                </div>
              </Collapse>
              <Collapse in={accountType.value === 'creditcard'}>
                <div>
                  <Input
                    type="select"
                    label="Credit Card Company"
                    {...accountCompany}
                  >
                    <option value="">select...</option>
                    <option value="Visa">Visa</option>
                    <option value="MasterCard">Mastercard</option>
                    <option value="American Express">American Express</option>
                    <option value="Discover">Discover</option>
                    <option value="Diners Club">Diners Club</option>
                    <option value="JCB">JCB</option>
                    <option value="Other">Other</option>
                  </Input>
                </div>
              </Collapse>
              <Button
                disabled={this.props.pristine}
                onClick={() => this.handleUpdateAccount(this.props.account)}
                bsStyle="success"
              >Update</Button>
              {' '}
              <Button
                onClick={this.toggleSettings}
              >Cancel</Button>
              <Button
                className="pull-right"
                disabled={this.state.confirmDeleteVisible}
                onClick={this.toggleConfirmDelete}
                bsStyle="danger"
              >Delete</Button>
              <div className="clearfix"></div>
              <Collapse in={this.state.confirmDeleteVisible}>
                <div>
                  <hr />
                  <p>Type DELETE into this box to confirm</p>
                  <div className="row">
                    <div className="col-xs-6">
                      <Input
                        type="text"
                        value={this.state.confirmDeleteText}
                        onChange={this.handleConfirmDeleteText}
                        placeholder="DELETE" />
                    </div>
                    <div className="col-xs-6">
                      <Button
                        disabled={this.state.confirmDeleteText !== 'DELETE'}
                        onClick={() => this.handleDeleteAccount(this.props.account)}
                        bsStyle="success"
                      >Confirm</Button>
                      {' '}
                      <Button
                        onClick={this.toggleConfirmDelete}
                      >Cancel</Button>
                    </div>
                  </div>
                </div>
              </Collapse>
            </form>
            { alertMessage }
          </div>
        </Collapse>
      </Well>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators({
    doToggleAccountDeletedConfirm: toggleAccountDeletedConfirm,
    doStoreDeletedAccountName: storeDeletedAccountName,
  }, dispatch);
}

export default reduxForm(
  {
    form: 'EditAccount',
    fields: ['accountName', 'accountType', 'accountCompany'],
  },
  null,
  mapDispatchToProps
)(AccountsListItem);
