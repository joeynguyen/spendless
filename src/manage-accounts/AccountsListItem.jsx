import PropTypes from 'prop-types';
import React, { Component } from 'react';
import EditAccountFormContainer from './EditAccountFormContainer.jsx';
import { Animate, Box } from 'grommet';

export default class AccountsListItem extends Component {
  static propTypes = {
    account: PropTypes.object.isRequired,
  }
  state = {
    settingsVisible: false,
  }
  toggleSettings = () => {
    this.setState({ settingsVisible: !this.state.settingsVisible });
  }
  render() {
    let editAccountForm = '';
    if (this.state.settingsVisible) {
      const formInitialValues = {
        initialValues: {
          accountName: this.props.account.name,
          accountType: this.props.account.type,
          accountCompany: this.props.account.company,
        }
      };
      // adding EditAccount form this way so that the component will mount and
      // unmount, resetting the form fields in doing so
      // also, this speeds up loading the modal because redux-form isn't initialized
      // for every account form on load. Only initializes when account is toggled
      editAccountForm = (
        <EditAccountFormContainer
          {...formInitialValues}
          account={this.props.account}
          formKey={this.props.account._id}
          toggleSettings={this.toggleSettings}
        />
      );
    }

    return (
      <Box
        colorIndex="accent-2-a"
        key={this.props.account._id}
        margin={{vertical: 'medium'}}
        pad="small"
      >
        <div className={(this.state.settingsVisible ? 'expanded' : '') + ' toggle-account-setting-icon'}>
          <i className="fa fa-lg fa-lg fa-fw fa-cog pull-right" onClick={this.toggleSettings} ></i>
        </div>
        <h4 className="account-name">{this.props.account.name}</h4>
        <p className="company-info">{this.props.account.type === 'bank' ? 'Bank' : 'Credit Card'} - {this.props.account.company}</p>
        <Animate
          enter={{animation: 'slide-down', duration: 300, delay: 0}}
          visible={this.state.settingsVisible}
        >
          { editAccountForm }
        </Animate>
      </Box>
    );
  }
}
