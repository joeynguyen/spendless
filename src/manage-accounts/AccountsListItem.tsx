import React, { Component } from 'react';
import EditAccountFormContainer from './EditAccountFormContainer.jsx';
import { Collapse } from 'react-collapse';
import { Icon } from 'antd';

type Props = {
  account: Account
}

type State = {
  settingsVisible: boolean,
}

export interface Account {
  _id: string;
  _rev: string;
  company: string;
  name: string;
  type: string;
}

export default class AccountsListItem extends Component<Props, State> {
  state = {
    settingsVisible: false,
  };
  toggleSettings = () => {
    this.setState({ settingsVisible: !this.state.settingsVisible });
  };
  render() {
    let editAccountForm : string | JSX.Element = '';
    if (this.state.settingsVisible) {
      const formInitialValues = {
        initialValues: {
          accountCompany: this.props.account.company,
          accountName: this.props.account.name,
          accountType: this.props.account.type,
        },
      };
      // adding EditAccount form this way so that the component will mount and
      // unmount, resetting the form fields in doing so
      // also, this speeds up loading the modal because forms aren't initialized
      // for every account form on load. Only initializes when account is toggled
      editAccountForm = (
        <EditAccountFormContainer
          {...formInitialValues}
          account={this.props.account}
          toggleSettings={this.toggleSettings}
        />
      );
    }

    const accountTypeText =
      this.props.account.type === 'bank' ? 'Bank' : 'Credit Card';
    return (
      <div className="well" key={this.props.account._id}>
        <div
          className={
            (this.state.settingsVisible ? 'expanded' : '') +
            ' toggle-account-setting-icon float-right'
          }
        >
          <Icon type="setting" onClick={this.toggleSettings} />
        </div>
        <h4 className="account-name">{this.props.account.name}</h4>
        <p className="company-info">
          {`${accountTypeText} - ${this.props.account.company}`}
        </p>
        <Collapse isOpened={this.state.settingsVisible}>
          {editAccountForm}
        </Collapse>
      </div>
    );
  }
}
