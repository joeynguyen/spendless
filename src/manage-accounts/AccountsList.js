import React, { PropTypes } from 'react';
import AccountsListItem from './AccountsListItem.js';

const AccountsList = ({ accounts }) => {
  return (
    <div>
    {
      accounts.map(function(account) {
        const formInitialValues = {
          initialValues: {
            accountName: account.name,
            accountType: account.type,
            accountCompany: account.company,
          }
        };
        return (
          <AccountsListItem {...formInitialValues} account={account} key={account._id} formKey={account._id} />
        );
      })
    }
    </div>
  );
};
AccountsList.propTypes = {
  accounts: PropTypes.array.isRequired,
};

export default AccountsList;
