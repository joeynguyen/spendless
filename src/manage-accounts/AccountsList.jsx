import PropTypes from 'prop-types';
import React from 'react';
import AccountsListItem from './AccountsListItem.jsx';

const AccountsList = ({ accounts }) => {
  return (
    <React.Fragment>
      {accounts.map(function(account) {
        return <AccountsListItem account={account} key={account._id} />;
      })}
    </React.Fragment>
  );
};
AccountsList.propTypes = {
  accounts: PropTypes.array.isRequired,
};

export default AccountsList;
