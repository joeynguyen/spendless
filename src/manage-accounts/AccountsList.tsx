import React from 'react';
import AccountsListItem, { Account } from './AccountsListItem';

type Props = {
  accounts: Account[];
}

const AccountsList = (props: Props) => {
  const accounts: Account[] = props.accounts;
  return (
    <>
      {accounts.map(function(account: Account) {
        return <AccountsListItem account={account} key={account._id} />;
      })}
    </>
  );
};

export default AccountsList;
