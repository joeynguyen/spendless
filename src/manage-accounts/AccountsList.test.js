import expect from 'expect';
import React from 'react';
import { shallow } from 'enzyme';
import AccountsList from './AccountsList';
import AccountsListItem from './AccountsListItem.js';

const myProps = {
  accounts: [
    {
      _id: '2016-04-24T20:47:37.374Z',
      _rev: '2-8c6e36eddfb9990c199546ad267cfd61',
      company: 'Santander Banks',
      name: 'Santander Checkings',
      type: 'bank'
    },
    {
      _id: '2016-02-24T07:01:24.948Z',
      _rev: '13-2bfa9371e269957dcbf2946c1cdd6cd5',
      company: 'MasterCard',
      name: 'Chase Freedom',
      type: 'creditcard'
    },
    {
      _id: '2016-03-24T08:10:58.232Z',
      _rev: '14-57184e4f5b50323eeeb6a919faf1af48',
      company: 'Bank of America',
      name: 'BofA Checkings',
      type: 'bank'
    }
  ]
};

function setup(props) {
  return shallow(<AccountsList accounts={props.accounts} />);
}

describe('AccountsList', () => {
  const wrapper = setup(myProps);
  const accountsListItems = wrapper.find('AccountsListItem');

  it('renders an AccountsListItem for each account', () => {
    expect(accountsListItems.length).toBe(3);
  });

  it('renders the correct props for each account', () => {
    myProps.accounts.forEach(( account, i ) => {
      expect(accountsListItems.at(i).prop('account')).toEqual(myProps.accounts[i]);
    });
  });
});
