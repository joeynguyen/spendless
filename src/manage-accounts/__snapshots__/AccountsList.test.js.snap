// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsList matches Jest snapshot 1`] = `
<Fragment>
  <AccountsListItem
    account={
      Object {
        "_id": "2016-04-24T20:47:37.374Z",
        "_rev": "2-8c6e36eddfb9990c199546ad267cfd61",
        "company": "Santander Banks",
        "name": "Santander Checkings",
        "type": "bank",
      }
    }
    key="2016-04-24T20:47:37.374Z"
  />
  <AccountsListItem
    account={
      Object {
        "_id": "2016-02-24T07:01:24.948Z",
        "_rev": "13-2bfa9371e269957dcbf2946c1cdd6cd5",
        "company": "MasterCard",
        "name": "Chase Freedom",
        "type": "creditcard",
      }
    }
    key="2016-02-24T07:01:24.948Z"
  />
  <AccountsListItem
    account={
      Object {
        "_id": "2016-03-24T08:10:58.232Z",
        "_rev": "14-57184e4f5b50323eeeb6a919faf1af48",
        "company": "Bank of America",
        "name": "BofA Checkings",
        "type": "bank",
      }
    }
    key="2016-03-24T08:10:58.232Z"
  />
</Fragment>
`;
